### PR TITLE
887 - Ethereum bytecode disassembler

### PR DESCRIPTION
**Pull request by molowny**
Disassembler for the Ethereum's contract bytecode. The code is based on the https://github.com/ethereum/evmdasm

### CODE REVIEW

File: conseil-common/src/test/scala/tech/cryptonomic/conseil/common/ethereum/domain/BytecodeTest.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package tech.cryptonomic.conseil.common.ethereum.domain
+
+import org.scalatest.wordspec.AnyWordSpec
+import org.scalatest.matchers.should.Matchers
+
+class BytecodeTest extends AnyWordSpec with Matchers {
+  "Bytecode" should {
+      "return opcodes form bytecode" in {
+        Bytecode("60406318160ddd00").opcodes shouldBe Seq(
```
[the remark this answers is not in the document]
Comment (molowny) [reply conseil-common/src/test/scala/tech/cryptonomic/conseil/common/ethereum/domain/BytecodeTest.scala:16]: This is an edge case with push instruction which requires to handle next bytes, so if it works for that then it's ok

File: conseil-common/src/test/scala/tech/cryptonomic/conseil/common/ethereum/domain/BytecodeTest.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          Opcode(0x2, Instructions.PUSH4, BigInt(0x18160ddd)),
+          Opcode(0x7, Instructions.STOP, BigInt(0x0))
+        )
+      }
+
+      "check if bytecode implements function" in {
+        Bytecode("60406318160ddd00").implements("totalSupply()") shouldBe true
+      }
+
+      "check if bytecode is a erc20 contract" in {
```
[the remark this answers is not in the document]
Comment (molowny) [reply conseil-common/src/test/scala/tech/cryptonomic/conseil/common/ethereum/domain/BytecodeTest.scala:28]: I've moved bytecodes to the fixtures, please check it now

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package tech.cryptonomic.conseil.common.ethereum
+
+import scorex.crypto.hash.Keccak256
+
+/**
+  * Utils methods required by the Ethereum integration.
+  */
+object Utils {
+  def functionSignatureTo4byteHexSelector(function: String): BigInt =
```
[the remark this answers is not in the document]
Comment (molowny) [reply conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala]: done

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala [hunk 4]
```diff
@@ -0,0 +1,81 @@
+package tech.cryptonomic.conseil.common.ethereum.domain
+
+import tech.cryptonomic.conseil.common.ethereum.Utils
+
+/**
+  * Ethereum contract bytecode disassembler.
+  * It allows to chech if given contract is ERC20 or ERC721.
+  * The code is based on https://github.com/ethereum/evmdasm
+  */
+case class Bytecode(bytecode: String) {
```
[the remark this answers is not in the document]
Comment (molowny) [reply conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala]: I've changed it to "value"

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+      }
+      ._2
+
+  /**
+    * Check if bytecode implements a particular function.
+    */
+  def implements(function: String): Boolean =
+    opcodes.exists(
+      o => o.instruction.name == "PUSH4" && o.parameters == Utils.functionSignatureTo4byteHexSelector(function)
```
[the remark this answers is not in the document]
Comment (molowny) [reply conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala]: done

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            case (skip, opcodes) => (skip - 1, opcodes)
+          }
+
+      }
+      ._2
+
+  /**
+    * Check if bytecode implements a particular function.
+    * Every function call is represented in the bytecode 
+    * by PUSH4 command with 4 bytes sigsignature.
```
Comment (ivanopagano) [inline conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala:61]: typo? sigsignature?

**Review by ivanopagano — APPROVED** [review]
LGTM

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala (hunk 6 above)
Comment (molowny) [reply conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/domain/Bytecode.scala]: fixed

**Review by ajuszczak — APPROVED** [review]

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+/**
+  * Utils methods required by the Ethereum integration.
+  */
+object Utils {
+
+  /**
+    * Normalize hex string by removing 0x prefix.
+    */
+  def remove0x(value: String): String =
+    value match {
```
Comment (piotrkosecki) [inline conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala:16]: I believe `value.stripPrefix("0x")` would work the same way.

Comment (molowny) [reply conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala]: good to know :) thanks!

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      case value if value.startsWith("0x") => value.substring(2)
+      case _ => value
+    }
+
+  /**
+    * Decode hex string.
+    */
+  def hexToString(value: String): String =
+    remove0x(value).sliding(2, 2).toArray.map(Integer.parseInt(_, 16).toChar).mkString.trim
```
Comment (piotrkosecki) [inline conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala:21]: I think `sliding(2,2)` can be replaced by `grouped(2)` like in `Bytecode:26`

**Review by piotrkosecki — APPROVED** [review]
Left two comments, but overall looks good!

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala [hunk 9]
```diff
@@ -12,11 +12,7 @@ object Utils {
   /**
     * Normalize hex string by removing 0x prefix.
     */
-  def remove0x(value: String): String =
-    value match {
-      case value if value.startsWith("0x") => value.substring(2)
-      case _ => value
-    }
+  def remove0x(value: String): String = value.stripPrefix("0x")
```
Comment (ivanopagano) [inline conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala:15]: I wonder if it makes sense anymore to keep this method instead of inlining the `stripPrefix`...

Comment (molowny) [reply conseil-common/src/main/scala/tech/cryptonomic/conseil/common/ethereum/Utils.scala:15]: I also have this impression, I will get rid of it
